Source infra

kofalt opened the pull request:
This PR:

* Removes sudo from the default launch target.
* Removes explicit dependency on apt from the default launch target.
* Adds documentation for installing system packages, with room for more distros / OSes.
* Compiles nginx & mongo from source for your system.
* Fixes a long-standing irritating with apt's nginx demanding non-root write access to `/var/log/nginx`.

At minimum, to merge we need:

* Confirmation this works on another Ubuntu x64 box
* Confirmation this works in vagrant
* Fix our CI failure to build mongo

Pinging @lmperry and @gsfr for review.

gsfr commented:
Worked for me as root on a fresh Ubuntu 15.04. Have to remember not to install into ```/root``` as that leads to permissions problems at runtime.
One minor suggestion/request:
- In config.toml, could we change the port names from "insecure" and "web" to "http" and "https" and the default http port to a more standard 8080?

kofalt commented:
Build fails for me on Travis and Vagrant. Mongo bug filed:

https://jira.mongodb.org/browse/CDRIVER-762

kofalt commented:
Ticket was closed, with a request to repeat on their mailing lists.
Suggestions welcome; I'll get around to shooting a list email at some point.

Eric89GXL commented:
@kofalt the wheels PR moves away from `sudo` use a little bit already. Maybe it's worth resurrecting this once the wheel PR is in or out? Compiling from source is a bit annoying, but should be cleaner than modifying people's `sources` and holding back packages, which the current setup seems to do. So +1 for me on this new approach. If your time is too limited to work on this, I can take it over after #61 is settled. Looks like you've done most of the hard work anyway :) so it will hopefully require a rebase plus some minor tweaks.

kofalt commented:
That sounds reasonable. However, you should know I'm lukewarm at best on this PR right now.

I was gung-ho about this initially, as I know from previous experience that nginx is trivial to build and takes ~30 seconds. Mongo, meanwhile... oh boy. Mongo takes  ~6 minutes on a modern quad-core laptop. This is completely unacceptable in say, a cloud environment which is likely dual-core at best and far slower to boot.

As an anecdote, it currently takes 17 minutes to install scitran on our dev server (a dual-core VM) without this branch. Landing wheels should drop that considerably, and I'd like to keep our install time as low as possible thereafter. 

Of the five benefits listed at the top of this thread, I'd like to keep 1, 2, 3, and 5.

I dunno when this changed (or maybe I'm blind) but there are [tarball binary downloads](https://www.mongodb.org/downloads) of mongo 3.x that we can pin. I think the new solution to this problem is:

* Upgrade everything to mongo 3.x, close #41 
* Based on the version in [mandantory.toml](https://github.com/scitran/scitran/blob/master/scripts/mandantory.toml) check if we have that version unpacked in `persistent/setup`.
* If not, download & unpack the tarball for your OS.
* Update reflex to use unpacked binary.
* Figure out solution for nginx. Maybe we leave the compilation stuff from this PR in. Maybe there's a better option. Ideally get away from using an embarrassing plaintext transport...

kofalt commented:
Incidentally, if we have to go completely crazy, we could consider using [fpm](https://github.com/jordansissel/fpm) which comes recommended by a colleague. Obviously I'd like to avoid that if possible.

kofalt commented:
I actually don't know if #41 is truly a blocker for this new proposal. I gave the changelogs a read and I think we could just switch major versions and nothing should break? Not a mongo expert.

Eric89GXL commented:
The new proposal sounds good to me. Compiling nginx doesn't sound like too bad an option for now, given that it is pretty quick.

gsfr commented:
I'm already on mongo 3.x locally, without any problems.

+1 for @kofalt's proposal, including nginx from source.

PR for scitran/api#41 is almost ready and expected to come out today.

kofalt commented:
Rebased. Was a little messy & required some cleanup.

@gsfr asserts that we can transparently upgrade to mongo 3.x, so I've removed the horrible compilation failure and replace it with binary tarballs distributed canonically.

Utter hilarity: Travis is too old to make SSL connections. Ref travis-ci/travis-ci#4757.
I, uh... I could write a five-line golang script to download a file?

kofalt commented:
Also of interest: `Disallowing packages: libatlas3gf-base, uuid-runtime`

@Eric89GXL you'd know more about `libatlas3gf-base` than I would, thoughts on needing it for the correct running (not building) of our python wheels? The container Travis option whitelists which apt packages they allow in an attempt to prevent container escapees, presumably. There is an [approval process](https://github.com/travis-ci/apt-package-whitelist#package-approval-process) we could attempt if necessary.

Eric89GXL commented:
Looking at what `scipy` does:

https://github.com/scipy/scipy/blob/master/.travis.yml

It looks like we can probably get away with installing `libatlas-dev libatlas-base-dev liblapack-dev`. But you could first try installing none of them and see if we get a linking problem when using our prebuilt libs.

Eric89GXL commented:
Looks like Travis is failing only because `EnsurePackages` is not found now

kofalt commented:
Fixed; ready for review & commentary.

I intend to leave the nginx solution in place for now. I'm not a fan of neither burning CPU time nor the plaintext transport, but it ~works~ and solves other problems we've been having. Will likely file a ticket to fix these problems after this lands.

Eric89GXL commented on .travis.yml:
```diff
@@ -9,6 +9,31 @@
+      - gfortran
+      - libatlas-dev
+      - libatlas-base-dev
+      - libatlas3gf-base
```
might as well remove since it doesn't get installed, maybe make a comment that it isn't installed because Travis doesn't allow it

kofalt replied:
Since it only results in a (hidden) warning, my inclination is to keep these packages identical in every way to what a user might need to install (including python compilation) on their system.

I'll add a comment to the top listing which are disabled / ignored.

Eric89GXL commented on .travis.yml:
```diff
@@ -9,6 +9,31 @@
+      - python-dev
+      - python-pip
+      - python-virtualenv
+      - uuid-runtime
```
same for this

Eric89GXL commented:
LGTM. After this I can remove the apt source that was added for mongo (I think?), right? This seems like a cleaner solution both for Travis and local installs.

kofalt commented:
Affirmative on removing the apt source. 

I'm happy to have all distro-specific setup confined to the optional `Prepare` target.
